Interactive Streaming Output Support

VasuBhog opened the pull request:
This PR introduces interactive streaming support by making changes to how we call the run command in the proxy. Iterating over the run command allows us to send information to the cell's output iteratively.

![PowerShell_streaming](https://user-images.githubusercontent.com/23587151/94325671-325da880-ff65-11ea-9e9b-47759b2edb16.gif)

VasuBhog committed:
Interactive Streaming via changing the run command

VasuBhog commented:
@chlafreniere, @pensivebrian, and @corivera  please review the PR. Looks like I am unable to assign individuals due to permissions. Let me know if you have any questions!

pensivebrian commented:
Anyway to add tests for this? In particular, really large strings to make sure we handle cases where a string being output by powershell over stdout is split over two messages.

pensivebrian commented on powershell_kernel/powershell_proxy.py:
```diff
@@ -40,8 +40,10 @@ def __init__(self, repl):
+        for temp in self.run_command('function prompt() {"^"}'):
+            continue
+        
     def run_command(self, input):
```
Let's add comment, something like:
`# returns a generator that yields string messages as the are returned from powershell via stdout`

pensivebrian reviewed (changes requested):
Change looks good. Waiting on tests and comments.

VasuBhog committed:
Addressed issue/comments as well as adding test for scenario

VasuBhog commented:
Addressed another issue when re-running a large string (over 10mb) after initial execution (2nd or 3rd time). Changing the write function to look for the **^** in a more efficient way. Added test using the jupyter kernel test tool for this specific scenario, did plenty of manual testing as well (large strings, and ^ within a string for example).

corivera commented on powershell_kernel/powershell_proxy.py:
```diff
@@ -58,11 +60,18 @@ def run_command(self, input):
-            self._repl.write(input + '\n')
-
+            self._repl.write(input + '\n') 
+            
```
Extra spaces here

corivera commented on powershell_kernel/powershell_proxy.py:
```diff
@@ -58,11 +60,18 @@ def run_command(self, input):
 
-            self._repl.write(input + '\n')
-
+            self._repl.write(input + '\n') 
```
Extra space at the end of this line

corivera commented on powershell_kernel/powershell_proxy.py:
```diff
@@ -92,8 +101,9 @@ def update_view_loop(self):
     def write(self, packet):
         # this is a hack to detect when we stop processing this input
-        if packet == '^':
+        if packet.endswith('^'):
```
Why endswith here? Is there a chance we might get actual output that ends in this character?

VasuBhog replied:
Yes, I used `endswith` due to the issue that currently exists when large packets need to be processed (for example a large string in a for loop) and the kernel hangs (and outputs ^ at the end) after the user tries to re-run the code cell multiple times. Since the long string example outputs the response by separating the message in different packets, it will cause the "^" to get attached to the end of the packet, as a result, using `endswith` solves the issue.

VasuBhog committed:
minor space change

corivera reviewed (approved):
Approved pending completion of the unit tests.

VasuBhog committed:
Added few more unit tests - working on utilizing other unit test methods to test the PS Kernel

VasuBhog committed:
PowerShell Kernel Baseline Notebook Test

VasuBhog committed:
Move ipynb test file to tests folder

VasuBhog commented:
This PR fixes #37
